File: src/main/java/com/raychenon/leetcode/ShortestCommonSupersequence.java
package com.raychenon.leetcode;

/**
 * User: raychenon
 * Date: 10/6/2020
 * https://leetcode.com/problems/shortest-common-supersequence/
 * <p>
 * https://en.wikipedia.org/wiki/Shortest_common_supersequence_problem
 */
public class ShortestCommonSupersequence {

    /**
     * time complexity : O(s1 * s2)
     * space complexity : O(s1 * s2)
     */
    public String shortestCommonSupersequence(String str1, String str2) {
        int n1 = str1.length();
        int n2 = str2.length();
        int[][] mem = new int[n1 + 1][n2 + 1];
        int i, j;

        for (i = 1; i <= n1; i++) {
            for (j = 1; j <= n2; j++) {
                char ch1 = str1.charAt(i - 1);
                char ch2 = str2.charAt(j - 1);
                if (ch1 == ch2) {
                    mem[i][j] = 1 + mem[i - 1][j - 1];
                } else {
                    mem[i][j] = Math.max(mem[i - 1][j], mem[i][j - 1]);
                }
            }
        }
        i = n1;
        j = n2;
        StringBuilder str = new StringBuilder();
        while (i > 0 && j > 0) {
            if (mem[i][j] == mem[i - 1][j]) {
                str.append(str1.charAt(i - 1));
                i--;
            } else if (mem[i][j] == mem[i][j - 1]) {
                str.append(str2.charAt(j - 1));
                j--;
            } else {
                str.append(str1.charAt(i - 1));
                i--;
                j--;
            }
        }
        while (i-- > 0) {
            str.append(str1.charAt(i));
        }
        while (j-- > 0) {
            str.append(str2.charAt(j));
        }
        return str.reverse().toString();
    }
}

File: src/main/kotlin/com/raychenon/kotlin/leetcode/ShortestCommonSupersequence.kt
package com.raychenon.kotlin.leetcode

import java.util.*


/**
 * User: raychenon
 * Date: 13/6/2020
 * https://leetcode.com/problems/shortest-common-supersequence/
 *
 * This problem is NP-hard
 */
object ShortestCommonSupersequence {

    fun shortestCommonSupersequence(str1: String, str2: String): String? {
        val lcs = longestCommonSubSeq(str1, str2)
        var p1 = 0
        var p2 = 0
        var sb = StringBuffer()
        for (i in 0 until lcs!!.length) {
            while (p1 < str1.length && str1[p1] != lcs[i]) {
                sb.append(str1[p1++])
            }
            while (p2 < str2.length && str2[p2] != lcs[i]) {
                sb.append(str2[p2++])
            }
            sb.append(lcs[i])
            p1++
            p2++
        }
        sb.append(str1.substring(p1)).append(str2.substring(p2))
        return sb.toString()
    }

    private fun longestCommonSubSeq(str1: String, str2: String): String? {
        val dp = Array(
            str1.length + 1
        ) { arrayOfNulls<String>(str2.length + 1) }
        for (i in dp.indices) {
            Arrays.fill(dp[i], "")
        }
        for (i in 1..str1.length) {
            for (j in 1..str2.length) {
                if (str1[i - 1] == str2[j - 1]) {
                    dp[i][j] = dp[i - 1][j - 1] + str1[i - 1]
                } else {
                    dp[i][j] =
                        if (dp[i - 1][j]!!.length > dp[i][j - 1]!!.length) dp[i - 1][j] else dp[i][j - 1]
                }
            }
        }
        return dp[str1.length][str2.length]
    }
}